src/mococrw/kdf.h

```diff
@@ -0,0 +1,117 @@
+/*
+ * #%L
+ * %%
+ * Copyright (C) 2020 BMW Car IT GmbH
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+#pragma once
+#include <memory>
+#include <vector>
+#include "hash.h"
+
+namespace mococrw
+{
+
+/**
+ * @brief Abstract class for key derivation functions
+ */
+class KeyDerivationFunction
+{
+public:
+    /**
+     * @brief destructor
+     */
+    virtual ~KeyDerivationFunction();
+    /**
+     * @brief Provides a derived key.
+     *
+     * This function returns a derived key based on the arguments to this function and the arguments provided to
+     * the constructor
+     * @param password The password/secret key which shall be used for key derivation
+     * @param outputLength The length of the derived key in bytes.
+     * @param salt The salt used for the key derivation
+     * @throws This method may throw an OpenSSLException if OpenSSL indicates an error.
+     * @return The derived key
+     */
+    virtual std::vector<uint8_t> deriveKey(const std::vector<uint8_t> &password, const size_t outputLength,
+                                           const std::vector<uint8_t> &salt) = 0;
+};
+
+class PBKDF2: public KeyDerivationFunction {
+public:
+    /**
+     * @brief Constructor
+     * @param hashFunction The hash function which shall be used
+     */
+    PBKDF2(openssl::DigestTypes hashFunction, uint32_t iterations);
+
+    ~PBKDF2();
+
+    /**
+     * @brief Provides a derived key.
+     *
+     * This function returns a derived key based on the arguments to this function and the arguments provided to
+     * the constructor
+     * @param password The password/secret key which shall be used for key derivation
+     * @param outputLength The length of the derived key in bytes.
+     * @param salt The salt used for the key derivation
+     * @throws This method may throw an OpenSSLException if OpenSSL indicates an error.
+     * @return The derived key
+     */
+    std::vector<uint8_t> deriveKey(const std::vector<uint8_t> &password, const size_t outputLength,
+                                   const std::vector<uint8_t> &salt) override;
+
+private:
+    openssl::DigestTypes hashFunction;
+    uint32_t iterations;
+    std::vector<uint8_t>_derivedKey;
```

Comment:
```suggestion
    std::vector<uint8_t> _derivedKey;
```

Also, why is this a member variable at all? I think it may be enough that we're returning this from `deriveKey()`.

Reply:
done